libuninameslist.3
.TH LIBUNINAMESLIST 3 "2003-Jan-16"
.SH NAME
libuninameslist \- Unicode Names and Annotations List
.SH SYNOPSIS
.B #include <uninameslist.h>
.br
Compile and link with
.I -luninameslist
.br
.SH DESCRIPTION
The
.B libuninameslist
library provides the ability for a program to fetch unicode
.B names
and
.B annotations
as described by www.unicode.org in a text file called
.B NamesList.txt
.SH OPTIONS
This library package comes with an optional Python wrapper
called
.I uninameslist.py
and an optional French nameslist library called
.I libuninameslist\-fr
.SH AUTHOR
Read
.I AUTHORS
file for everyone involved in making
.B libuninameslist


ChangeLog
- 2017-Aug-07
    * Version 1.2, Unicode 10.0, French version is now also 10.0.0.
    * Added function calls to French version to allow use of one or more
      libraries at the same (Dynamic library linking needs testing since
      added 'FR' may be issue with older programs expecting same name).
    * Added more "make check" to include checking French library.

- 2017-Jul-01
    * Version 1.1, Unicode 10.0, French version is 7.0(2014-06-22).
    * Added "make check" to allow for automatic library test and verify.

- 2017-Mar-19
    * Unicode 10.0_Beta (Nameslist.txt 10_Beta for review until Apr30).

- 2016-Jul-01
    * Version 1.0, Unicode 9.0, French version is 7.0(2014-06-22).
    * Corrected library to report correctly as next version up. This is
      probably a significant change, therefore bumped both libraries to
      start at 1.0 even though backwards compatibility remains the same,
      otherwise would have become version_0.6 (and French version_0.4).
    * Removed the excessively long package name, which you'll note as a
      so-bump change with Fedora and also Debian in naming this package.

- 2015-Jul-01
    * Version 0.5. Unicode 8.0. French version is 7.0(2014-06-22).

- 2014-Jul-31
    * Version 0.4. Unicode 7.0. Updated French version to 7.0(2014-06-22).

- 2014-Mar-08
    * Unicode 7.0_Beta (Nameslist.txt 7.0 for review, French stays at 5.0).
    * Added function for block {start, end, name, number} for version 0.4.
    * Cleaned up several warnings for program buildnameslist.

- 2013-May-01
    * Makefile.am problems in changing nameslist.h to uninameslist.h, and
      nameslist-fr.h to uninameslist-fr.h, therefore renamed those files.
    * Removed several files that can be built during autoreconf and automake
      steps to building a tar.gz package.
    * Added additional check for -enable-oldfrenchlib to "make distcheck"
    * This will be tagged as version 0.3

- 2013-Apr-28
    * Separate namelist.h and namelist-fr.h due to different versions.
    * Clarify README to indicated differences between both libraries.
    * Closes bugs http://sourceforge.net "WFLAGS value - ID: 1307144",
	"pkg-config support and fix for bug 1307144 - ID: 3402865"
	"Please update autoconf - ID: 3608906",
    * http://sourceforge.net/projects/libuninameslist/ not being maintained.

- 2013-Apr-23
    * Convert build system to autotools method (configure.ac, Makefile.am)
    * Create README. Add function to report NamesList Version (6.2).
    * Prep files for a distribution tar ball - version 0.3.20130423

- 2013-Feb-19
    * Unicode 6.3 (Nameslist.txt remains at 6.2, French remains at 5.0)

- 2012-Aug-22
    * Unicode 6.1

- 2009-Dec-31
    * Unicode 5.2

- 2008-Apr-07
    * Unicode 5.1
      (Also the French translation has been updated to 5.0).

- 2006-Sep-07
    * Unicode 5.0

- 2005-Jul-12
    * Unicode 4.1

- 2003-Jul-13
    * Oops the format of the NamesList.txt file is slightly different and I was inserting too many newlines.

- 2003-Jul-10
    * Upgrade to Unicode 4.0

- 2003-Jan-16
    * Initial release


libuninameslist.py
# libuninameslist
#
# Copyright (C) 2017, Shriramana Sharma
#
# This Python wrapper is subject to the same "BSD 3-clause"-type license
# which the wrapped C library is subject to.

__all__ = ["version", "name", "annotation", "blocks", "block", "uplus"]

# connecting to the dynamic library

from ctypes import *
_lib = CDLL("libuninameslist.so.1")

def _setSig(fn, restype, argtypes):
    if restype is not None: fn.restype = restype
    fn.argtypes = argtypes

# const char *uniNamesList_NamesListVersion(void);
_setSig(_lib.uniNamesList_NamesListVersion, c_char_p, [])
# const char *uniNamesList_name(unsigned long uni);
_setSig(_lib.uniNamesList_name, c_char_p, [c_ulong])
# const char *uniNamesList_annot(unsigned long uni);
_setSig(_lib.uniNamesList_annot, c_char_p, [c_ulong])
# int uniNamesList_blockCount(void);
_setSig(_lib.uniNamesList_blockCount, c_int, [])
# int uniNamesList_blockNumber(unsigned long uni);
_setSig(_lib.uniNamesList_blockNumber, c_int, [c_ulong])
# long uniNamesList_blockStart(int uniBlock);
_setSig(_lib.uniNamesList_blockStart, c_long, [c_int])
# long uniNamesList_blockEnd(int uniBlock);
_setSig(_lib.uniNamesList_blockEnd, c_long, [c_int])
# const char *uniNamesList_blockName(int uniBlock);
_setSig(_lib.uniNamesList_blockName, c_char_p, [c_int])

# internal helpers

class _block:

    __slots__ = ["name", "start", "end"]

    def __init__(self, name, start, end):
        self.name = name; self.start = start; self.end = end

    def __str__(self):
        return "<‘{}’: {} - {}>".format(self.name, uplus(self.start), uplus(self.end))

    @staticmethod
    def fromNum(num):
        return _block(_lib.uniNamesList_blockName(num).decode(),
                      _lib.uniNamesList_blockStart(num),
                      _lib.uniNamesList_blockEnd(num))

_blockCount = _lib.uniNamesList_blockCount()

# public symbols

version = _lib.uniNamesList_NamesListVersion().decode()

def name(char):
    return _lib.uniNamesList_name(ord(char)).decode()

def annotation(char):
    annot = _lib.uniNamesList_annot(ord(char))
    return "" if annot is None else annot.decode()

def blocks():
    for blockNum in range(_blockCount):
        yield _block.fromNum(blockNum)

def block(char):
    return _block.fromNum(_lib.uniNamesList_blockNumber(ord(char)))

def uplus(char):
    if type(char) is int:
        if not (0 <= char <= 0x10FFFF):
            raise ValueError("Invalid Unicode codepoint: U+{:X}".format(char))
        val = char
    else:
        val = ord(char)
    return ("U+{:06X}" if val > 0xFFFF else "U+{:04X}").format(val)

# test

def _test():
    print("Using libuninameslist version :", version)
    print("The Unicode name of 😃 is      :", name("😃"))
    print("The Unicode annotation of 😃 is:")
    print(annotation("😃"))
    print("The Unicode block of 😃 is     :", block("😃"))
    print("The Unicode codepoint of 😃 is :", uplus("😃"))
    print()
    print("A complete list of blocks:")
    print("\n".join(str(block) for block in blocks()))

if __name__ == "__main__":
    _test()
